python/lsst/pipe/base/quantum_graph/aggregator/_progress.py

```diff
@@ -0,0 +1,209 @@
+    logging when progress bars for interactive display are in use.  The context
+    manager does nothing otherwise.
+    """
+
+    def __init__(self, log: LsstLogAdapter, config: AggregatorConfig):
+        self.start = time.time()
+        self.log = log
+        self.config = config
+        self._periodic_log = PeriodicLogger(self.log, config.log_status_interval)
+        self._n_scanned: int = 0
+        self._n_ingested: int = 0
+        self._n_written: int = 0
+        self._n_quanta: int | None = None
+        self.interactive = config.interactive_status
+
+    def __enter__(self) -> Self:
+        if self.interactive:
+            from tqdm.contrib.logging import logging_redirect_tqdm
+
+            self._logging_redirect = logging_redirect_tqdm()
+            self._logging_redirect.__enter__()
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        traceback: TracebackType | None,
+    ) -> bool | None:
+        if self.interactive:
+            self._logging_redirect.__exit__(exc_type, exc_value, traceback)
+        return None
+
+    def set_n_quanta(self, n_quanta: int) -> None:
+        """Set the total number of quanta.
+
+        Parameters
+        ----------
+        n_quanta : `int`
+            Total number of quanta, including special "init" quanta.
+
+        Notes
+        -----
+        This method must be called before any of the ``report_*`` methods or
+        `finish_resuming`.
+        """
+        self._n_quanta = n_quanta
+        if self.interactive:
+            from tqdm import tqdm
+
+            self._scan_progress = tqdm(desc="Scanning", total=n_quanta, leave=False, unit="quanta")
+            self._ingest_progress = tqdm(
+                desc="Ingesting", total=n_quanta, leave=False, smoothing=0.1, unit="quanta"
+            )
+            if self.config.output_path is not None:
+                self._write_progress = tqdm(desc="Writing", total=n_quanta, leave=False, unit="quanta")
+
+    @property
+    def elapsed_time(self) -> float:
+        """The time in seconds since the start of the aggregator."""
+        return time.time() - self.start
+
+    def _log_status(self) -> None:
+        """Invoke the periodic logger with the current status."""
+        self._periodic_log.log(
+            "%s quanta scanned, %s quantum outputs ingested, "
+            "%s provenance quanta written (of %s) after %0.1fs.",
+            self._n_scanned,
+            self._n_ingested,
+            self._n_written,
+            self._n_quanta,
+            self.elapsed_time,
+        )
+
+    def report_scan(self) -> None:
+        """Report that a quantum was scanned."""
+        self._n_scanned += 1
+        if self.interactive:
+            self._scan_progress.update(1)
+        else:
+            self._log_status()
+
+    def finish_scans(self) -> None:
```

Comment:
I would have named it `report_finished_scans` or `report_scans_done` to match other function as well as to not sound like triggering something to finish the scans.  Same for the ingests and writes.